In [1]:
from parlay_system import *

# Creating a MoneyLine object
A moneyline is represented as either a positive or a negative number

### Positive MoneyLine
Positive moneyline returns represent the underdog if the odds are +165 that means that you bet 100 to win  165 for a total return of 265. This can be calculated as follows:

#### Calculating profit

$ profit = BET * \frac{ODDS}{100} $

$ profit = 100 * \frac{165}{100} $

$ profit = 165 $

#### Calculating payout

$ payout = BET + profit $

$ payout = 100 + 165 $

$ payout = 265 $

The **MoneyLine()** class takes an event name, bet amount, and odds and will calculate this information for us

In [2]:
pos_ml = MoneyLine(event="positive", bet_amount=100, odds=165)
pos_ml.print_stats()

      event  bet_amount  odds  multiplier  payout
0  positive         100   165        2.65   265.0


### Negative MoneyLine
Negative moneyline returns represent the **favorite** if the odds are -150 that means that you bet 150 to win 100 for a total return of 166. This can be calculated as follows:

#### Calculating profit

$ profit = BET * \bigl(\frac{100}{ODDS} \bigr) $

$ profit = 100 * \bigl(\frac{100}{150} \bigr) $

$ profit = 66.67 $

#### Calculating payout

$ payout = BET + profit $

$ payout = 100 + 66.67 $

$ payout = 166.67 $

In [3]:
neg_ml = MoneyLine(event="negative", bet_amount=100, odds=-150)
neg_ml.print_stats()

      event  bet_amount  odds  multiplier      payout
0  negative         100  -150    1.666667  166.666667


# Creating a Parlay object
A parlay object combines the multipliers from two moneyline objects to create an even hgher return (with more risk)

Here is how you can calculate a "True Odds" Parlay

Let's say we want a parlay with the following teams:
pacers              -150
celtics             +170
bucks               -120


First, we have to determine what the multipliers would be for each game, simply divide what the total payout would be (risk + win) by the risk amount

In [4]:
pacers = MoneyLine(event="pacers", bet_amount=100, odds=-150)
celtics = MoneyLine(event="celtics", bet_amount=100, odds=170)
bucks = MoneyLine(event="bucks", bet_amount=100, odds=-120)

In [5]:
pacers.print_stats()

    event  bet_amount  odds  multiplier      payout
0  pacers         100  -150    1.666667  166.666667


In [6]:
celtics.print_stats()

     event  bet_amount  odds  multiplier  payout
0  celtics         100   170         2.7   270.0


In [7]:
bucks.print_stats()

   event  bet_amount  odds  multiplier      payout
0  bucks         100  -120    1.833333  183.333333


### Parlay Formula
So to create a parlay for these three events. You simply multiple the three multipliers by each other:

$ {parlay\_multiplier} = pacers_{multiplier} * celtics_{multiplier} * bucks_{multiplier} $


$ {parlay\_multiplier} = 1.67 * 2.7 * 1.83 $

$ {parlay\_multiplier} = 8.35 $ 

Now we will use the **Parlay()** class to create a parlay from these three moneylines

In [8]:
parlay = Parlay(money_line_arr=[pacers, celtics, bucks], bet_amount=100)

In [15]:
# Note: This result will be slightly off due to rounding error
parlay.print_stats()

                  event  bet_amount   odds  multiplier  payout
0  pacers_celtics_bucks         100  725.0        8.25   825.0


# Checking our results with Draftkings
These two classes are necessary for the next step which is to build a system that maximizes return while minimizing risk, by leveraging a combination of parlays. The last step us double chacking that the values calculated by these classes match the values that are giving on draftkings. 

In [10]:
bulls = MoneyLine(event="bulls", bet_amount=100, odds=200)
mavericks = MoneyLine(event="mavericks", bet_amount=100, odds=525)
rockets = MoneyLine(event="rockets", bet_amount=100, odds=-560)

In [11]:
bulls.print_stats()

   event  bet_amount  odds  multiplier  payout
0  bulls         100   200         3.0   300.0


![title](img/one_team_parlay.png)

In [12]:
two_team_parlay = Parlay(
    money_line_arr=[bulls, mavericks], 
    bet_amount=100)

two_team_parlay.print_stats()

             event  bet_amount    odds  multiplier  payout
0  bulls_mavericks         100  1775.0       18.75  1875.0


![title](img/two_team_parlay.png)

In [13]:
three_team_parlay = Parlay(
    money_line_arr=[bulls, mavericks, rockets], 
    bet_amount=100)

three_team_parlay.print_stats()

                     event  bet_amount         odds  multiplier       payout
0  bulls_mavericks_rockets         100  2109.821429   22.098214  2209.821429


![title](img/three_team_parlay.png)

### Notes on the discrpancy
As you can see, the Parlay() class came up with a result of 2209.82 while draftkings calculated a result of 2212.50 for the same parlay. This is likely due to a rounding error as they probably have more fractional values for each moneyline. For the purposes of this analysis well assume that they are the same. The minute differences shouldn't make a differece